### ISSUE TITLE
[Bug]: matshow skips norm intervals

### ISSUE THREAD

**Issue by daniprec** [matplotlib/matplotlib#24350]
### Bug summary

`matplotlib.pyplot.matshow` assigns values to the wrong `norm` interval.

### Code for reproduction

```python
import matplotlib.colors as clr
import matplotlib.pyplot as plt
import numpy as np

# Define the colormap and boundaries
cmap = clr.ListedColormap(["k", "b", "cyan", "g", "y", "r"])
boundaries = [0, 1, 2, 3, 6, 10, 63]
norm = clr.BoundaryNorm(boundaries, cmap.N)

# First example: matrix of size 2x16
# Values range from 0 to 15
mat = np.tile(np.arange(0, 16), (2, 1))
# First column is 0 so it should be black (interval 0-1)
# Second column is 1 so it should be blue (interval 1-2)
# Third column is 2 so it should be cyan (interval 2-3)
# Fourth column is 3 so it should be green (interval 3-6)
# Seventh column is 6 so it should be yellow (interval 6-10)
# Eleventh column is 10 so it should be red (interval 10-63)
plt.matshow(mat, cmap=cmap, norm=norm)
plt.title("Example 1: Range(0, 15)")
plt.show()
# Output is as expected

# Second example: matrix of size 2x26
# Values range from 0 to 26
mat = np.tile(np.arange(0, 26), (2, 1))
# Same expected ouput as before
plt.matshow(mat, cmap=cmap, norm=norm)
plt.title("Example 3: Range(0, 25) - Norm before matshow")
plt.show()
# Second column is 1, but it is painted black instead of blue

# Third example: same matrix 2x26
# We normalize it first before inputting into matshow
plt.matshow(norm(mat), cmap=cmap)
plt.title("Example 3: Range(0, 25) - Norm before matshow")
plt.show()
# Output is as expected
```

### Actual outcome

![imagen](https://user-images.githubusercontent.com/34949631/199710999-71974cda-b92e-477d-a780-f95c4429c6c5.png)
![imagen](https://user-images.githubusercontent.com/34949631/199711009-cee24357-0c3e-4806-9e97-15614bcdeeca.png)
![imagen](https://user-images.githubusercontent.com/34949631/199711028-f35e9193-3556-44ea-ac7a-5c1665ccbd54.png)

### Expected outcome

The expect outcome for **Example 2** should be the one seen in **Example 3**.

### Additional information

The same bug happened when ussing version 3.4 and 3.6. I have not tried in any other version.

### Operating system

Ubuntu

### Matplotlib Version

3.6.2

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.13

### Installation

conda

**Comment by jklymak** [comment]
I think you are just hitting a floating point instability because your data are right on the boundary of the norm intervals.  If you want this to be stable your norm intervals should be offset from your potential data points.

**Comment by daniprec** [comment]
@jklymak Thanks for your reply! I also thought that the problem was caused by the matrix values being exactly the boundary values. After testing `matplotlib.colors.BoundaryNorm` I saw that it always takes the intervals opened to the right. If value Y falls between intervals [X, Y) and [Y, Z), it assigns the color of the former. That is what happens in examples 1 and 3 above.

However in example 2, the value 1 is assigned the color of the left interval, while all the other values (that also fall in boundaries) are assigned the interval to the right.

**Comment by jklymak** [comment]
> However in example 2, the value 1 is assigned the color of the left interval, while all the other values (that also fall in boundaries) are assigned the interval to the right.

Yes, a == a will fall "randomly" on the right or left.   (of course it's not really random, but its floating-point implementation dependent).  We can't do anything about this, and the end user needs to supply intervals that are not likely to fall on data values.
